### PR TITLE
Various slice-handling ergonomic improvements

### LINKED ISSUE #47: Change 'add_assign' to take a byte slice rather than another 'SecretKey'
Hi, 

I encountered this limitation a few times for the use case of threshold and multi signatures. 
In practice tweaks are more likely to be hashes (or zero) than things that are conceptually "secret keys". 
Other elliptic curve libraries seems to allow this tweak and doing field arithmetic using zero element.

### PR DESCRIPTION
**Pull request by apoelstra**
Fixes #47

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -337,13 +345,26 @@ impl Message {
     pub fn from_slice(data: &[u8]) -> Result<Message, Error> {
         match data.len() {
             constants::MESSAGE_SIZE => {
+                let secp = Secp256k1::without_caps();
```
Comment (TheBlueMatt) [inline src/lib.rs]: Hmm, I know this is supposed to become free once upstream has a static context, but I'd like to at least see concept ACKs on that happening upstream before we merge something that relies on it.

File: CHANGELOG.md [hunk 2]
```diff
@@ -2,6 +2,8 @@
 # [Unreleased]
 
 * Update `rand` to 0.4 and `gcc` 0.3 to `cc` 1.0. (`rand` 0.5 exists but has a lot of breaking changes and no longer compiles with 1.14.0.)
+* Disallow messages that are 0 or overflow the curve order; this should never happen (probability < 2^128) with a random hash, and 0 in particular is cryptographically dangerous.
```
Comment (TheBlueMatt) [inline CHANGELOG.md]: Technically you can still create a bad Message via the From<[u8; 32]> impl, which really sucks, tbh. Can we convert that to a non-trait from function that returns a Result? I know its an API change which sucks but we might as well wait for the static context from upstream to make the from_slice() check free anyway?

File: src/key.rs [hunk 3]
```diff
@@ -96,11 +96,17 @@ impl SecretKey {
...
     #[inline]
     /// Adds one secret key to another, modulo the curve order
-    pub fn add_assign<C>(&mut self, secp: &Secp256k1<C>, other: &SecretKey)
-                     -> Result<(), Error> {
+    pub fn add_assign<C>(
+        &mut self,
+        secp: &Secp256k1<C>,
+        other: &[u8],
+    ) -> Result<(), Error> {
+        if other.len() != 32 {
```
Comment (TheBlueMatt) [inline src/key.rs]: Can you document this precondition?

File: src/lib.rs [hunk 4]
```diff
@@ -205,6 +205,14 @@ fn from_str(s: &str) -> Result<Signature, Error> {
 #[derive(Copy, Clone, PartialEq, Eq, Debug)]
 pub struct RecoverableSignature(ffi::RecoverableSignature);
 
+/// Trait describing something that promises to be a 32-byte random number; in particular,
+/// it has negligible probability of being zero or overflowing the group order. Such objects
+/// may be converted to `Message`s without any error paths.
```
Comment (real-or-random) [inline src/lib.rs]: cool stuff

**Pull request merged**